File: Backend/persistence/repositories/userRepository.ts
import { supabase } from '../../config/supabaseClient'; // Cliente Supabase configurado
import { User } from '../models/User';

class UserRepository {
  async create(userData: Partial<User>): Promise<User> {
    const { data, error } = await supabase
      .from('users') // 'users' é o nome da sua tabela no Supabase
      .insert([userData])
      .select()
      .single(); // .single() para retornar um único objeto ou null

    if (error) {
      console.error('Erro ao criar usuário no Supabase:', error);
      throw new Error('Não foi possível criar o usuário.');
    }
    if (!data) {
        throw new Error('Nenhum dado retornado após a criação do usuário.');
    }
    return data as User;
  }

  async findByEmail(email: string): Promise<User | null> {
    const { data, error } = await supabase
      .from('users')
      .select('*')
      .eq('email', email)
      .maybeSingle(); // .maybeSingle() retorna um único objeto ou null se não encontrar

    if (error) {
      console.error('Erro ao buscar usuário por email:', error);
      // Dependendo da sua política de erros, você pode querer lançar o erro ou apenas retornar null
    }
    return data ? (data as User) : null;
  }
  // Outros métodos: findById, update, delete, etc.
}
export const userRepository = new UserRepository();

File: Backend/persistence/repositories/user_repository.py
from config.database import supabase
from application.dtos.create_user_dto import CreateUserDTO
from datetime import date, datetime

class UserRepository:
    def get_user_by_email(self, email: str) -> dict | None:
        """
        Busca um usuário pelo email no banco de dados.
        Retorna um dicionário se o usuário for encontrado, ou None caso contrário.
        """
        response = supabase.table('usuarios').select('*').eq('email', email).execute()
        if response.data:
            return response.data[0]
        return None

    def add_user(self, user_data: CreateUserDTO, hashed_senha: str) -> dict:
        """
        Adiciona um novo usuário ao banco de dados.
        Retorna um dicionário com os dados do usuário criado.
        """
        user_dict = user_data.dict()
        user_dict['senha'] = hashed_senha
        
        for key, value in user_dict.items():
            if isinstance(value, (date, datetime)):
                user_dict[key] = value.isoformat()

        response = supabase.table('usuarios').insert(user_dict).execute()
        
        if response.data:
            return response.data[0]
        return None
    
    def get_all(self):
        """ Busca todos os usuários da tabela 'usuarios'."""
        try:
            response = supabase.table('usuarios').select('id', 'nome', 'email', 'senha', 'telefone', 'cpf', 'data_nascimento','tipo_usuario','criado_em').execute()
            return response.data
        except Exception as e:
            print(f"Erro ao buscar usuários: {e}")
            return None